Add blog. Add initial drafts for first 2 posts (#133)
infinum/eightshift-docs: website/blog/2022-01-21-components-and-blocks.md

---
title: Components and blocks
description: Explanation of the difference between components and blocks on a practical example
slug: components-and-blocks
authors: iobrado
tags: [eightshift, boilerplate, components, blocks]
image: #
hide_table_of_contents: false
---
In this blog post, we'll give you a practical explanation about the difference between components and blocks. For a truly in-depth explanation about components and blocks, you can refer to [our documentation](https://infinum.github.io/eightshift-docs/docs/basics/blocks). More specific explanation about component in a block can be found in [Component in a Block](https://infinum.github.io/eightshift-docs/docs/basics/blocks-component-in-block/) section.

<!--truncate-->
One of the things that is really confusing to newcomers is the difference between components and blocks. Let's break it all down.

### Component

> **A component is like a blank slate, a template without context**. It doesn't know or even care for what it will be used. Can you combine multiple simple components into one complex component? Absolutely! That's exactly what we'll do in the following example. Let's say we have a component which consists of an image and a heading.  Image and heading are two simple components that come with the theme out-of-the-box.

Reviewer: ```suggestion
**A component is like a blank slate, a template without context**. It doesn't know or even cares for what it will be used. Can you combine multiple simple components into one complex component? Absolutely! That's exactly what we'll do in the following example. 

Let's say we have a component that consists of an image and a heading. Image and heading are two simple components that come with the theme out-of-the-box.
```

> Here are a few examples how this new component can be used:

Reviewer: ```suggestion
Here are a few examples of how this new component can be used:
```

- displaying a post
- displaying a user or an author
- displaying a photo with the with the photographer's name
- and many more...

> As you can already see, there are many ways you can use this component. What helps even more to expand the usability of this component are `attributes`. Lets say we have the following attributes:

Reviewer: ```suggestion
As you can already see, there are many ways you can use this component. What helps even more to expand the usability of this component are `attributes`. Let's say we have the following attributes:
```

- `imagePosition` - left, right, top, bottom
- `imageShape` - rectangular, circular
- `imageWidth` - 25%, 50%, 75%
- `headingColor` - black, blue, red
- `headingSize` - small, medium, large

These attributes give you the option to modify the look of the component, so this specific component can be used for the following:

- **related posts** - rectangular image on the left side, 25% image width, heading takes up the remaining 75%, medium heading colored in blue
- **user avatar** - circular image on the left side, 25% image width, small heading coloured in black
- **photo of the day** - rectangular image on the right side which takes 75% width and photographer's name is written in medium blue text

> - **featured posts** - rectangular image on top, large heading in red color

Reviewer: Images of these would be awesome!

Author: Soon™ 😄

You even have the option to add conditional logic, e.g. to have the option for `imageWidth` only when `imagePosition` is set to left or right. And when `imagePosition` is set to top or bottom, it defaults to 100% width.

### Block

> So, what about blocks? Blocks can consist of one or more components and, most importantly, allow you to use them in editor.

Reviewer: ```suggestion
So, what about blocks? Blocks can consist of one or more components and, most importantly, allow you to use them in the editor.
```

> **Blocks are what gives components context.**

Reviewer: ```suggestion
**Blocks are what give components context.**
```

> Now that we have our component, let's call it `card-simple`, we can use it in a `Featured Posts` block. We already have predefined attributes how our component should look when used in this block defined in manifest.json, but we also need another attribute which will allow us to select which posts will be displayed.

Reviewer: ```suggestion
Now that we have our component, let's call it `card-simple`, we can use it in a `Featured Posts` block. We already have predefined attributes of how our component should look when used in this block defined in `manifest.json`, but we also need another attribute that will allow us to select which posts will be displayed.
```

> In order to do that, we add a new attribute called `featuredPosts` which defines which posts will be displayed in our block. For each selected post, another `card-simple` component is added and featured image and post title are dynamically added to its image and heading components. More about the query logic and adding new options in a block will be covered in a future chapter.

Reviewer: ```suggestion
In order to do that, we add a new attribute called `featuredPosts` which defines which posts will be displayed in our block. For each selected post, another `card-simple` component is added and featured image and post title are dynamically added to its image and heading components. More about the query logic and adding new options in a block will be covered in a future post.
```

> You may be wondering, can you have a block which allows you to use all options of the `card-simple` component? Of course! You can create a block which allows you to set all available options in editor and manually add content to image and heading. There are also block variations, which allow you to set up predefined attributes for a block. More about setting up block variations may be covered in one of the future chapters.

Reviewer: ```suggestion
You may be wondering, can you have a block that allows you to use all options of the `card-simple` component? Of course! You can create a block that allows you to set all available options in the editor and manually add content to the image and heading. There are also block variations, which allow you to set up predefined attributes for a block. More about setting up block variations may be covered in one of the future chapters.
```

### Conclusion

> Hopefully these examples helped you get a better understanding of the difference between components and blocks. It may take a bit more time or practical use to truly understand the difference and the reason behind it. Our next chapter will walk you through downloading one of our components and blocks with WP CLI and modifying it.

Reviewer: ```suggestion
Hopefully, these examples helped you get a better understanding of the difference between components and blocks. It may take a bit more time or practical use to understand the difference and reasoning behind it. Our next chapter will walk you through downloading one of our components and blocks with WP CLI and modifying it.
```

---
title: Components and blocks
description: Explanation of the difference between components and blocks on a practical example
slug: components-and-blocks
authors: iobrado
tags: [eightshift, boilerplate, components, blocks]
image: #
hide_table_of_contents: false
---

> In this blog post, we'll give you a practical explanation about the difference between components and blocks. For a truly in-depth explanation about components and blocks, you can refer to [our documentation](https://infinum.github.io/eightshift-docs/docs/basics/blocks). A more specific explanation about component in a block can be found in [Component in a Block](https://infinum.github.io/eightshift-docs/docs/basics/blocks-component-in-block/) section.

Reviewer 2: ```suggestion
In this blog post, we'll give you a practical explanation about the difference between components and blocks. For a truly in-depth explanation about components and blocks, you can refer to [our documentation](https://infinum.github.io/eightshift-docs/docs/basics/blocks). A more specific explanation about using components in a block can be found in the [Component in a Block](https://infinum.github.io/eightshift-docs/docs/basics/blocks-component-in-block/) section.
```

<!--truncate-->
One of the things that is really confusing to newcomers is the difference between components and blocks. Let's break it all down.

### Component

> **A component is like a blank slate, a template without context**. It doesn't know or even cares for what it will be used. Can you combine multiple simple components into one complex component? Absolutely! That's exactly what we'll do in the following example. 

Reviewer 2: ```suggestion
**A component is like a blank slate, a template without context**. It doesn't know or even care for what it will be used. Can you combine multiple simple components into one complex component? Absolutely! That's exactly what we'll do in the following example. 
```

Let's say we have a component that consists of an image and a heading. Image and heading are two simple components that come with the theme out-of-the-box.

Here are a few examples of how this new component can be used:
- displaying a post
- displaying a user or an author
- displaying a photo with the photographer's name
- and many more...

> As you can already see, there are many ways you can use this component. What helps even more to expand the usability of this component are `attributes`. Let's say we have the following attributes:

Reviewer 2: ```suggestion
As you can already see, there are many ways you can use this component. What helps expand the usability of this component even more are `attributes`. Let's say we have the following attributes:
```

- `imagePosition` - left, right, top, bottom
- `imageShape` - rectangular, circular
- `imageWidth` - 25%, 50%, 75%
- `headingColor` - black, blue, red
- `headingSize` - small, medium, large

These attributes give you the option to modify the look of the component, so this specific component can be used for the following:

- **related posts** - rectangular image on the left side, 25% image width, heading takes up the remaining 75%, medium heading colored in blue
- **user avatar** - circular image on the left side, 25% image width, small heading colored in black
- **photo of the day** - rectangular image on the right side which takes 75% width and photographer's name is written in medium blue text
- **featured posts** - rectangular image on top, large heading in red color

> You even have the option to add conditional logic, e.g. to have the option for `imageWidth` only when `imagePosition` is set to left or right. And when `imagePosition` is set to top or bottom, it defaults to 100% width.

Reviewer 2: Not sure what you're trying to say here - that you can conditionally hide or show options in Gutenberg?

### Block

> So, what about blocks? Blocks can consist of one or more components and, most importantly, allow you to use them in the editor.

Reviewer 2: Blocks can consist of zero components, technically 😄

Author: Yes, they can. I'll try to add that part, but I don't want to overcomplicate things 😄

**Blocks are what give components context.**

> Now that we have our component, let's call it `card-simple`, we can use it in a `Featured Posts` block. We already have predefined attributes of how our component should look when used in this block defined in `manifest.json`, but we also need another attribute that will allow us to select which posts will be displayed.

Reviewer 2: ```suggestion
Now that we have our component - let's call it `card-simple` - we can use it in a `Featured Posts` block. We already have predefined attributes of how our component should look when used in this block defined in `manifest.json`, but we also need another attribute that will allow us to select which posts will be displayed.
```

Not sure what you're trying to say with "We already have predefined attributes of how our component should look when used in this block defined in `manifest.json`".

> To do that, we add a new attribute called `featuredPosts` which defines which posts will be displayed in our block. For each selected post, another `card-simple` component is added and featured image and post title are dynamically added to its image and heading components. More about the query logic and adding new options in a block will be covered in a future post.

Reviewer 2: ```suggestion
To do that, we'll add a new attribute called `featuredPosts` which will define which posts will be displayed in our block. For each selected post, another `card-simple` component is added, and featured image and post title are dynamically added to its image and heading components. More about the query logic and adding new options in a block will be covered in a future post.
```
